### PR TITLE
Fix inconsistencies with error handling

### PR DESCRIPTION
**Pull request by sergejostir**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

I finally got around to fixing issues exposed in https://github.com/fastify/fastify/issues/2860. This is a a draft PR to indicate that I am actively working on these issues. It's not finished yet because I got issues creating tests that have to deal with uncaught exceptions and unhandled rejections (ref https://github.com/tapjs/node-tap/issues/797). Fastify doesn't really have any tests which would target the issues that I outlined in the beforementioned issue, therefore I will add a whole new test suite.

The agenda that I am following is that the process should crash when there is an error that cannot be sensibly handled. For example, in Fastify, that means when an error happens after the reply was already sent and there's nothing Fastify can do any more. This aproach matches the goals outlined in https://www.fastify.io/docs/latest/Reference/Errors/ and believe also follows the generally accepted best practices.

### CODE REVIEW

File: lib/handleRequest.js [hunk 1]
```diff
@@ -8,7 +8,11 @@ const {
 } = require('./symbols')
 
 function handleRequest (err, request, reply) {
-  if (reply.sent === true) return
+  if (reply.sent === true) {
```
Comment (climba03003) [inline lib/handleRequest.js:11]: If you throw the error like this. What happen if people call the `reply` in hook?

**Comment by mcollina** [comment]
I find this behaviour quite confusing because it does not match the behavior of promises, consider the following:

```js
const p = await new Promise((resolve, reject) => {
  resolve(42)
  reject(new Error('kaboom'))
})
```

I do not like this pattern, however it's how our platform work.
I'm not sure we should migrate away from it and crash our applications in that case. Wdyt @fastify/collaborators?
I think we might want to make sure those errors are logged and not swallowed.

---

You can test about `uncaughtException` and `unhandledRejection` by doing:

```js
const listeners = process.listeners(event)
process.removeListeners(event)
t.teardown(() => {
  for (const listener of listeners) {
    process.on(event, listener)
  }
})
```

**Comment by sergiz** [comment]
Logging errors is the same as silently swallowing them; you can act upon them programically. The errors have to be thrown, so you either fix your application or properly handle them.

**Comment by sergiz** [comment]
> ```js
> const listeners = process.listeners(event)
> process.removeListeners(event)
> t.teardown(() => {
>   for (const listener of listeners) {
>     process.on(event, listener)
>   }
> })
> ```

Any example? What's the value of  the`event`?

**Comment by mcollina** [comment]
uncaughtException or unhandledRejection

**Comment by sergiz** [comment]
That probably worked back in the day when tap added listeners for those events, but that is not the case any more.

**Pull request closed without merging**